File: python/quaternion.py
import math

class Quaternion:
    def __init__(self, w, x, y, z):
        self.w = w
        self.x = x
        self.y = y
        self.z = z

    def __repr__(self):
        return f"Quaternion(w={self.w}, x={self.x}, y={self.y}, z={self.z})"

def mult(q1, q2): # multiply 2 quaternions
    q3 = Quaternion(0, 0, 0, 0)
    q3.w = q1.w * q2.w - q1.x * q2.x - q1.y * q2.y - q1.z * q2.z
    q3.x = q1.w * q2.x + q1.x * q2.w + q1.y * q2.z - q1.z * q2.y
    q3.y = q1.w * q2.y - q1.x * q2.z + q1.y * q2.w + q1.z * q2.x
    q3.z = q1.w * q2.z + q1.x * q2.y - q1.y * q2.x + q1.z * q2.w
    return q3

def axis_angle_to_quaternion(axis, angle): # expects axis as a unit vector

    half_angle = angle / 2

    return Quaternion(
        math.cos(half_angle),
        axis[0] * math.sin(half_angle),
        axis[1] * math.sin(half_angle),
        axis[2] * math.sin(half_angle)
    )

def conj(q): # conjugate
    return Quaternion(q.w, -q.x, -q.y, -q.z)

def rotate_vector(q, v): # v' = qvq*; v expected as a 3d vector [x,y,z]
    v_q = Quaternion(0, v[0], v[1], v[2])
    return mult(mult(q,v_q), conj(q))

def mag(v): # return magnitude of a 3d vector
    return math.sqrt(v[0]**2 + v[1]**2 + v[2]**2)

def normalise(v): # normalise a 3d vector
    magnitude = mag(v)

    if magnitude == 0: return v

    vn = [0, 0, 0]

    for i in range(len(v)):
        vn[i] = v[i] / magnitude
    return vn

def quaternion_to_euler(q): ### CONVENTIONS FOR THIS PROJECT : Rotation about X = yaw, rotation about Y = roll, rotation about Z = pitch
    yaw = -math.atan2(
        2 * (q.w * q.x + q.y * q.z),
        1 - 2 * (q.x**2 + q.y**2)
    )

    value = 2 * (q.w * q.y - q.z * q.x)
    value = max(-1.0, min(1.0, value))

    roll = math.asin(value)

    pitch = -math.atan2(
        2 * (q.w * q.z + q.x * q.y),
        1 - 2 * (q.y**2 + q.z**2)
    )

    return [
        math.degrees(pitch),
        math.degrees(yaw),
        math.degrees(roll)
    ]

File: Core/Src/quaternion.c
#include "quaternion.h"
#include <math.h>

Quaternion quaternion_multiply(Quaternion q1, Quaternion q2) {
    Quaternion q3 = {
        q1.w * q2.w - q1.x * q2.x - q1.y * q2.y - q1.z * q2.z,
        q1.w * q2.x + q1.x * q2.w + q1.y * q2.z - q1.z * q2.y,
        q1.w * q2.y - q1.x * q2.z + q1.y * q2.w + q1.z * q2.x,
        q1.w * q2.z + q1.x * q2.y - q1.y * q2.x + q1.z * q2.w,
    };

    return q3;
}

// Takes an axis as a unit vector and an angle and constructs a quaternion
Quaternion quaternion_from_axis_angle(Vector3 axis, float angle) {
    angle /= 2;

    Quaternion q = {
        cosf(angle),
        axis.x * sinf(angle),
        axis.y * sinf(angle),
        axis.z * sinf(angle)
    };

    return q;
}

Quaternion quaternion_conjugate(Quaternion q) {
    Quaternion conjugate = {
        q.w,
        -q.x,
        -q.y,
        -q.z
    };

    return conjugate;
}

// applies a rotation, q, to a vector, v
Vector3 quaternion_apply_rotation(Quaternion q, Vector3 v) {
    Quaternion v_quaternion = {
        0.0f,
        v.x,
        v.y,
        v.z
    };

    Quaternion result = quaternion_multiply(
        quaternion_multiply(q,v_quaternion),
        quaternion_conjugate(q)
    );

    Vector3 rotated = {
        result.x,
        result.y,
        result.z
    };

    return rotated;
}

// converts a quaternion to a Vector3 of euler angles - pitch, yaw, roll
// conventions for this project:
// Rotation about X = yaw
// Rotation about Y = roll
// Rotation about Z = pitch
Vector3 quaternion_to_euler(Quaternion q) {
    float yaw = -atan2f(
        2 * (q.w * q.z + q.x * q.y),
        1 - 2 * (q.x*q.x + q.y*q.y)
    );

    float value = 2 * (q.w * q.y - q.z * q.x);
    value = fmaxf(-1.0f, fminf(1.0f, value));

    float roll = asinf(value);

    float pitch = -atan2f(
        2 * (q.w * q.z + q.x * q.y),
        1 - 2 * (q.y*q.y + q.z*q.z)
    );

    Vector3 euler = {
        pitch,
        yaw,
        roll
    };

    return euler;
}